Use Vector::uncheckedAppend() less in WebCore

cdumez opened the pull request:
#### ba50d1bc34e9629af103d7bcdaf825f47b4cfc5e
<pre>
Use Vector::uncheckedAppend() less in WebCore
<a href="https://bugs.webkit.org/show_bug.cgi?id=262832">https://bugs.webkit.org/show_bug.cgi?id=262832</a>

Reviewed by Darin Adler.

Use Vector::uncheckedAppend() less in WebCore and use more efficient alternatives
now that uncheckedAppend() has become an alias for append().

* Source/WebCore/Modules/streams/ReadableStream.cpp:
(WebCore::ReadableStream::tee):
* Source/WebCore/Modules/webxr/WebXRHand.cpp:
(WebCore::WebXRHand::WebXRHand):
* Source/WebCore/PAL/pal/avfoundation/OutputContext.mm:
(PAL::OutputContext::outputDevices const):
* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::objectsForIDs const):
* Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
(WebCore::labelsForNode):
(WebCore::labelForNode):
(WebCore::AccessibilityNodeObject::checkboxOrRadioRect const):
(WebCore::AccessibilityNodeObject::correspondingLabelForControlElement const):
(WebCore::AccessibilityNodeObject::textForLabelElements const):
(WebCore::AccessibilityNodeObject::titleUIElement const):
(): Deleted.
* Source/WebCore/accessibility/AccessibilityNodeObject.h:
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.cpp:
(WebCore::AXIsolatedObject::children):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.cpp:
(WebCore::AXIsolatedTree::resolveAppends):
* Source/WebCore/animation/CSSPropertyAnimation.cpp:
(WebCore::CSSPropertyAnimationWrapperMap::CSSPropertyAnimationWrapperMap):
* Source/WebCore/contentextensions/ContentExtensionCompiler.cpp:
(WebCore::ContentExtensions::addUniversalActionsToDFA):
* Source/WebCore/contentextensions/DFAMinimizer.cpp:
(WebCore::ContentExtensions::DFAMinimizer::minimize):
* Source/WebCore/contentextensions/DFANode.cpp:
(WebCore::ContentExtensions::DFANode::actions const):
* Source/WebCore/css/CSSValueList.cpp:
(WebCore::CSSValueContainingVector::copyValues const):
* Source/WebCore/css/CSSVariableData.cpp:
(WebCore::CSSVariableData::CSSVariableData):
* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::ComputedStyleExtractor::copyProperties const):
* Source/WebCore/css/StyleProperties.cpp:
(WebCore::StyleProperties::copyProperties const):
* Source/WebCore/css/parser/CSSParserTokenRange.h:
(WebCore::CSSParserTokenRange::size const):
* Source/WebCore/css/typedom/HashMapStylePropertyMapReadOnly.cpp:
(WebCore::HashMapStylePropertyMapReadOnly::entries const):
* Source/WebCore/css/typedom/StylePropertyMapReadOnly.cpp:
(WebCore::StylePropertyMapReadOnly::reifyValueToVector):
* Source/WebCore/cssjit/SelectorCompiler.cpp:
(WebCore::SelectorCompiler::SelectorCodeGenerator::generateElementIsNthChild):
(WebCore::SelectorCompiler::SelectorCodeGenerator::generateElementIsNthLastChild):
* Source/WebCore/dom/ComposedTreeIterator.h:
(WebCore::ComposedTreeIterator::ComposedTreeIterator):
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::getAttributeNames const):
* Source/WebCore/dom/ElementData.h:
(WebCore::AttributeConstIterator::operator--):
(WebCore::AttributeIteratorAccessor::size const):
* Source/WebCore/dom/ImageOverlay.cpp:
(WebCore::ImageOverlay::updateWithTextRecognitionResult):
* Source/WebCore/dom/SelectorQuery.cpp:
(WebCore::SelectorDataList::SelectorDataList):
* Source/WebCore/dom/TouchList.h:
(WebCore::TouchList::TouchList):
* Source/WebCore/html/FileInputType.cpp:
(WebCore::FileInputType::filesChosen):
* Source/WebCore/html/HTMLFormElement.cpp:
(WebCore::HTMLFormElement::textFieldValues const):
* Source/WebCore/html/LinkIconCollector.cpp:
(WebCore::LinkIconCollector::iconsOfTypes):
* Source/WebCore/loader/DocumentLoader.cpp:
(WebCore::DocumentLoader::setActiveContentRuleListActionPatterns):
* Source/WebCore/loader/cache/CachedRawResource.cpp:
(WebCore::CachedRawResource::didAddClient):
* Source/WebCore/page/Page.cpp:
(WebCore::Page::replaceRangesWithText):
* Source/WebCore/page/mac/ServicesOverlayController.mm:
(WebCore::ServicesOverlayController::buildSelectionHighlight):
* Source/WebCore/platform/FileChooser.cpp:
(WebCore::FileChooser::chooseFiles):
* Source/WebCore/platform/SharedBuffer.cpp:
(WebCore::FragmentedSharedBuffer::copy const):
* Source/WebCore/platform/animation/AcceleratedEffectValues.cpp:
(WebCore::AcceleratedEffectValues::AcceleratedEffectValues):
* Source/WebCore/platform/audio/AudioBus.cpp:
(WebCore::AudioBus::AudioBus):
* Source/WebCore/platform/audio/AudioDSPKernelProcessor.cpp:
(WebCore::AudioDSPKernelProcessor::initialize):
* Source/WebCore/platform/audio/AudioResampler.cpp:
(WebCore::AudioResampler::AudioResampler):
* Source/WebCore/platform/audio/DynamicsCompressorKernel.cpp:
(WebCore::DynamicsCompressorKernel::setNumberOfChannels):
* Source/WebCore/platform/audio/MultiChannelResampler.cpp:
(WebCore::MultiChannelResampler::MultiChannelResampler):
* Source/WebCore/platform/graphics/ComplexTextController.cpp:
(WebCore::ComplexTextController::finishConstruction):
* Source/WebCore/platform/graphics/FontCache.cpp:
(WebCore::FontCache::purgeInactiveFontData):
* Source/WebCore/platform/graphics/FontCascade.cpp:
(WebCore::FontCascade::characterSelectionRectsForText const):
* Source/WebCore/platform/graphics/FontCascadeCache.cpp:
(WebCore::makeFontCascadeCacheKey):
* Source/WebCore/platform/graphics/GraphicsLayer.h:
(WebCore::KeyframeValueList::KeyframeValueList):
* Source/WebCore/platform/graphics/PathUtilities.cpp:
(WebCore::PathUtilities::pathsWithShrinkWrappedRects):
* Source/WebCore/platform/graphics/SourceBufferPrivate.cpp:
(WebCore::SourceBufferPrivate::trackBuffersRanges const):
* Source/WebCore/platform/graphics/avfoundation/CDMFairPlayStreaming.cpp:
(WebCore::extractSinfData):
* Source/WebCore/platform/graphics/avfoundation/objc/CDMInstanceFairPlayStreamingAVFObjC.mm:
(WebCore::CDMInstanceSessionFairPlayStreamingAVFObjC::copyKeyStatuses const):

Canonical link: <a href="https://commits.webkit.org/269062@main">https://commits.webkit.org/269062@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/19c226ba410cdf581b2277632b165ab15dd272bf

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/21504 "2 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/21807 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/22557 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/23368 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/19929 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/21746 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/25113 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/22058 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/21122 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/21731 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/21366 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/18644 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/24219 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/18558 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/19494 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/25806 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/19645 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/19701 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/23655 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/20184 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/17193 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/19501 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/5129 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/23741 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/20089 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/contentextensions/ContentExtensionCompiler.cpp:
```diff
@@ -182,8 +182,7 @@ static void addUniversalActionsToDFA(DFA& dfa, UniversalActionSet&& universalAct
     dfa.actions.reserveCapacity(dfa.actions.size() + universalActions.size());
-    for (uint64_t action : universalActions)
-        dfa.actions.uncheckedAppend(action);
+    dfa.actions.appendRange(universalActions.begin(), universalActions.end());
```
Do we benefit from doing `reserveCapacity` before `appendRange?`

cdumez replied:
At the moment, yes. However, I think we should consider optimizing Vector::appendRange() and reserve capacity at least when the input iterators are random access iterators (and std::distance() can return the size in constant time). I can look into this but likely in a follow-up.

cdumez replied:
Following up in https://github.com/WebKit/WebKit/pull/18839.

darinadler commented on Source/WebCore/css/CSSValueList.cpp:
```diff
@@ -211,11 +211,9 @@ bool CSSValueContainingVector::hasValue(CSSValueID otherValue) const
-    for (auto& value : *this)
-        builder.uncheckedAppend(const_cast<CSSValue&>(value));
-    return builder;
+    return WTF::map<4>(*this, [](auto& value) -> Ref<CSSValue> {
```
The idea is that `CSSValueListBuilder` could change its inline capacity in the future. The hardcoded 4 here doesn’t seem great, especially if it will make working but inefficient code in case of mismatch. I suppose it’s OK if compilation will just fail.

cdumez replied:
It fails to build at the moment:
```
Source/WebCore/css/CSSValueList.cpp:214:12: error: no viable conversion from returned value of type 'Vector<[...], 0UL aka 0, [2 * ...]>' to function return type 'Vector<[...], 4, [2 * ...]>'
```

That said, it is probably better to use a global constant to be safe. Nothing stops someones from updating the Vector class to convert silently.

darinadler commented on Source/WebCore/dom/ImageOverlay.cpp:
```diff
@@ -645,16 +642,18 @@ void updateWithTextRecognitionResult(HTMLElement& element, const TextRecognition
-        auto& block = result.blocks[index];
+
+    size_t index = 0;
+    Vector<FontSizeAdjustmentState> elementsToAdjust = WTF::compactMap(result.blocks, [&](auto& block) -> std::optional<FontSizeAdjustmentState> {
```
`auto`?

darinadler commented on Source/WebCore/loader/DocumentLoader.cpp:
```diff
@@ -2596,14 +2596,12 @@ void DocumentLoader::setActiveContentRuleListActionPatterns(const HashMap<String
-        Vector<UserContentURLPattern> patternVector;
-        patternVector.reserveInitialCapacity(pair.value.size());
-        for (auto& patternString : pair.value) {
+        Vector<UserContentURLPattern> patternVector = WTF::compactMap(pair.value, [](auto& patternString) -> std::optional<UserContentURLPattern> {
```
`auto`?

darinadler commented on Source/WebCore/page/mac/ServicesOverlayController.mm:
```diff
@@ -401,14 +401,11 @@ static void compactRectsWithGapRects(Vector<LayoutRect>& rects, const Vector<Gap
-        cgRects.reserveInitialCapacity(m_currentSelectionRects.size());
-
-        for (auto& rect : m_currentSelectionRects) {
+        Vector<CGRect> cgRects = WTF::map(m_currentSelectionRects, [&](auto& rect) -> CGRect {
```
`auto`?

cdumez committed:
Use Vector::uncheckedAppend() less in WebCore
https://bugs.webkit.org/show_bug.cgi?id=262832

Reviewed by Darin Adler.

Use Vector::uncheckedAppend() less in WebCore and use more efficient alternatives
now that uncheckedAppend() has become an alias for append().

* Source/WebCore/Modules/streams/ReadableStream.cpp:
(WebCore::ReadableStream::tee):
* Source/WebCore/Modules/webxr/WebXRHand.cpp:
(WebCore::WebXRHand::WebXRHand):
* Source/WebCore/PAL/pal/avfoundation/OutputContext.mm:
(PAL::OutputContext::outputDevices const):
* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::objectsForIDs const):
* Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
(WebCore::labelsForNode):
(WebCore::labelForNode):
(WebCore::AccessibilityNodeObject::checkboxOrRadioRect const):
(WebCore::AccessibilityNodeObject::correspondingLabelForControlElement const):
(WebCore::AccessibilityNodeObject::textForLabelElements const):
(WebCore::AccessibilityNodeObject::titleUIElement const):
(): Deleted.
* Source/WebCore/accessibility/AccessibilityNodeObject.h:
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.cpp:
(WebCore::AXIsolatedObject::children):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.cpp:
(WebCore::AXIsolatedTree::resolveAppends):
* Source/WebCore/animation/CSSPropertyAnimation.cpp:
(WebCore::CSSPropertyAnimationWrapperMap::CSSPropertyAnimationWrapperMap):
* Source/WebCore/contentextensions/ContentExtensionCompiler.cpp:
(WebCore::ContentExtensions::addUniversalActionsToDFA):
* Source/WebCore/contentextensions/DFAMinimizer.cpp:
(WebCore::ContentExtensions::DFAMinimizer::minimize):
* Source/WebCore/contentextensions/DFANode.cpp:
(WebCore::ContentExtensions::DFANode::actions const):
* Source/WebCore/css/CSSValueList.cpp:
(WebCore::CSSValueContainingVector::copyValues const):
* Source/WebCore/css/CSSVariableData.cpp:
(WebCore::CSSVariableData::CSSVariableData):
* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::ComputedStyleExtractor::copyProperties const):
* Source/WebCore/css/StyleProperties.cpp:
(WebCore::StyleProperties::copyProperties const):
* Source/WebCore/css/parser/CSSParserTokenRange.h:
(WebCore::CSSParserTokenRange::size const):
* Source/WebCore/css/typedom/HashMapStylePropertyMapReadOnly.cpp:
(WebCore::HashMapStylePropertyMapReadOnly::entries const):
* Source/WebCore/css/typedom/StylePropertyMapReadOnly.cpp:
(WebCore::StylePropertyMapReadOnly::reifyValueToVector):
* Source/WebCore/cssjit/SelectorCompiler.cpp:
(WebCore::SelectorCompiler::SelectorCodeGenerator::generateElementIsNthChild):
(WebCore::SelectorCompiler::SelectorCodeGenerator::generateElementIsNthLastChild):
* Source/WebCore/dom/ComposedTreeIterator.h:
(WebCore::ComposedTreeIterator::ComposedTreeIterator):
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::getAttributeNames const):
* Source/WebCore/dom/ElementData.h:
(WebCore::AttributeConstIterator::operator--):
(WebCore::AttributeIteratorAccessor::size const):
* Source/WebCore/dom/ImageOverlay.cpp:
(WebCore::ImageOverlay::updateWithTextRecognitionResult):
* Source/WebCore/dom/SelectorQuery.cpp:
(WebCore::SelectorDataList::SelectorDataList):
* Source/WebCore/dom/TouchList.h:
(WebCore::TouchList::TouchList):
* Source/WebCore/html/FileInputType.cpp:
(WebCore::FileInputType::filesChosen):
* Source/WebCore/html/HTMLFormElement.cpp:
(WebCore::HTMLFormElement::textFieldValues const):
* Source/WebCore/html/LinkIconCollector.cpp:
(WebCore::LinkIconCollector::iconsOfTypes):
* Source/WebCore/loader/DocumentLoader.cpp:
(WebCore::DocumentLoader::setActiveContentRuleListActionPatterns):
* Source/WebCore/loader/cache/CachedRawResource.cpp:
(WebCore::CachedRawResource::didAddClient):
* Source/WebCore/page/Page.cpp:
(WebCore::Page::replaceRangesWithText):
* Source/WebCore/page/mac/ServicesOverlayController.mm:
(WebCore::ServicesOverlayController::buildSelectionHighlight):
* Source/WebCore/platform/FileChooser.cpp:
(WebCore::FileChooser::chooseFiles):
* Source/WebCore/platform/SharedBuffer.cpp:
(WebCore::FragmentedSharedBuffer::copy const):
* Source/WebCore/platform/animation/AcceleratedEffectValues.cpp:
(WebCore::AcceleratedEffectValues::AcceleratedEffectValues):
* Source/WebCore/platform/audio/AudioBus.cpp:
(WebCore::AudioBus::AudioBus):
* Source/WebCore/platform/audio/AudioDSPKernelProcessor.cpp:
(WebCore::AudioDSPKernelProcessor::initialize):
* Source/WebCore/platform/audio/AudioResampler.cpp:
(WebCore::AudioResampler::AudioResampler):
* Source/WebCore/platform/audio/DynamicsCompressorKernel.cpp:
(WebCore::DynamicsCompressorKernel::setNumberOfChannels):
* Source/WebCore/platform/audio/MultiChannelResampler.cpp:
(WebCore::MultiChannelResampler::MultiChannelResampler):
* Source/WebCore/platform/graphics/ComplexTextController.cpp:
(WebCore::ComplexTextController::finishConstruction):
* Source/WebCore/platform/graphics/FontCache.cpp:
(WebCore::FontCache::purgeInactiveFontData):
* Source/WebCore/platform/graphics/FontCascade.cpp:
(WebCore::FontCascade::characterSelectionRectsForText const):
* Source/WebCore/platform/graphics/FontCascadeCache.cpp:
(WebCore::makeFontCascadeCacheKey):
* Source/WebCore/platform/graphics/GraphicsLayer.h:
(WebCore::KeyframeValueList::KeyframeValueList):
* Source/WebCore/platform/graphics/PathUtilities.cpp:
(WebCore::PathUtilities::pathsWithShrinkWrappedRects):
* Source/WebCore/platform/graphics/SourceBufferPrivate.cpp:
(WebCore::SourceBufferPrivate::trackBuffersRanges const):
* Source/WebCore/platform/graphics/avfoundation/CDMFairPlayStreaming.cpp:
(WebCore::extractSinfData):
* Source/WebCore/platform/graphics/avfoundation/objc/CDMInstanceFairPlayStreamingAVFObjC.mm:
(WebCore::CDMInstanceSessionFairPlayStreamingAVFObjC::copyKeyStatuses const):

Canonical link: https://commits.webkit.org/269062@main
